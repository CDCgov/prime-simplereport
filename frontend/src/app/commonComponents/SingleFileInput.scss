.sr-single-file-input {
  position: relative;
  max-width: 30rem;

  input[type="file"] {
    cursor: pointer;
    border: 0.06rem dashed #a9aeb1;
    padding: 3.5rem 1rem 2rem 7.2rem;
    z-index: 2;
    position: relative;
    display: block;
    width: 100%;
    margin: 0;
    font-size: 0.88rem;
  }

  input::file-selector-button {
    border: none;
    cursor: pointer;
    background: transparent;
    text-decoration: underline;
    color: #005ea2;
    font-weight: normal;
  }

  span {
    position: absolute;
    top: 0;
    z-index: 1;
    width: 100%;
    font-size: 0.88rem;
    text-align: center;
    padding: 2rem 1rem 2rem 1rem;
  }
}

.sr-single-file-input.selected {
  background-color: #d9e8f6;

  input[type="file"] {
    font-weight: bold;
    padding-left: 4.4rem;

    &::after {
      content: "\2611";
      margin-left: 0.25rem;
    }
  }
}

.sr-single-file-input.invalid {
  background-color: transparent;

  span {
    font-weight: bold;
    color: #b50909;
  }

  input[type="file"] {
    padding-left: 9.1rem;

    &::after {
      content: "\2612";
      margin-left: 0.25rem;
      color: #b50909;
    }
  }
}

@media screen and (max-width: 457px) {
  .sr-single-file-input input[type="file"],
  .sr-single-file-input.invalid input[type="file"],
  .sr-single-file-input.selected input[type="file"] {
    padding-left: 2rem;
  }
}
