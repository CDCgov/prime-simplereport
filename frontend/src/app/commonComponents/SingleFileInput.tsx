import React, { ChangeEventHandler, useState } from "react";
import classnames from "classnames";

import "./SingleFileInput.scss";

type SingleFileInputProps = {
  id: string;
  name: string;
  ariaLabel: string;
  accept?: string;
  onChange: ChangeEventHandler<HTMLInputElement>;
  required: boolean;
};

type FileState = "blank" | "invalid" | "selected";

const SingleFileInput = ({
  id,
  name,
  onChange,
  ariaLabel,
  accept,
  required = false,
}: SingleFileInputProps) => {
  const [fileState, setFileState] = useState<FileState>("blank");

  function handleOnChange(e: React.ChangeEvent<HTMLInputElement>) {
    if (e?.currentTarget?.files?.length === 0) {
      setFileState("blank");
    } else {
      const file = e?.currentTarget?.files?.item(0);

      if (accept && file?.type && isInvalidFile(file?.type, accept)) {
        setFileState("invalid");
      } else {
        setFileState("selected");
      }
    }
    onChange(e);
  }

  function getHint(fileState: FileState): string {
    switch (fileState) {
      case "invalid":
        return "This is not a valid file type";
      case "selected":
        return "Drag file here or choose from folder to change file";
      default:
        return "Drag file here or choose from folder";
    }
  }

  return (
    <div className={classnames(["sr-single-file-input", fileState])}>
      <input
        type="file"
        id={id}
        data-testid={id}
        name={name}
        required={required}
        onChange={handleOnChange}
        aria-label={ariaLabel}
        className="usa-file-input"
        aria-describedby="file-input-specific-hint"
        accept={accept}
      />
      <span id="file-input-specific-hint">{getHint(fileState)}</span>
    </div>
  );
};

function isInvalidFile(type: string, accept: string): boolean {
  const trimmedAcceptProp = accept.replace(/^\./, "").replace(/\/\*$/, "");

  if (
    type.search(new RegExp(trimmedAcceptProp)) !== -1 ||
    trimmedAcceptProp.search(new RegExp(type)) !== -1
  ) {
    return false;
  }

  return true;
}

export default SingleFileInput;
