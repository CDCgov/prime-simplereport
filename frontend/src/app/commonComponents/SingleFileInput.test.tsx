import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SingleFileInput from "./SingleFileInput";

const file = (text: BlobPart) => {
  const blob = new Blob([text]);
  const file = new File([blob], "values.csv", { type: "text/csv" });
  return file;
};

describe("Single File Input", () => {
  const handleOnChange = jest.fn();

  /**
   * const validFileContents =
   *   "Patient_last_name,Patient_first_name,Patient_middle_name,Patient_suffix,Patient_tribal_affiliation,Patient_ID,Ordered_test_code,Specimen_source_site_code,Specimen_type_code,Device_ID,Instrument_ID,Result_ID,Corrected_result_ID,Test_correction_reason,Test_result_status,Test_result_code,Illness_onset_date,Specimen_collection_date_time,Order_test_date,Test_date,Date_result_released,Patient_race,Patient_DOB,Patient_gender,Patient_ethnicity,Patient_preferred_language,Patient_street,Patient_street_2,Patient_city,Patient_state,Patient_zip_code,Patient_country,Patient_phone_number,Patient_county,Patient_email,Patient_role,Processing_mode_code,Employed_in_healthcare,Resident_congregate_setting,First_test,Symptomatic_for_disease,Testing_lab_name,Testing_lab_CLIA,Testing_lab_street,Testing_lab_street_2,Testing_lab_city,Testing_lab_state,Testing_lab_zip_code,Testing_lab_phone_number,Testing_lab_county,Organization_name,Ordering_facility_name,Ordering_facility_street,Ordering_facility_street_2,Ordering_facility_city,Ordering_facility_state,Ordering_facility_zip_code,Ordering_facility_phone_number,Ordering_facility_county,Ordering_provider_ID,Ordering_provider_last_name,Ordering_provider_first_name,Ordering_provider_street,Ordering_provider_street_2,Ordering_provider_city,Ordering_provider_state,Ordering_provider_zip_code,Ordering_provider_phone_number,Ordering_provider_county,Site_of_care\n" +
   *   "Alaska1,Judy,Suellen,,39,xen4p,94558-4,71836000,440500007,BD Veritor System for Rapid Detection of SARS-CoV-2,939273,Alaska1,613603,nh1rigems,F,260373001,20220225,202202201809-0500,202202261540-0500,20220221,20220220,2106-3,,F,H,Iloko,10269 Larry Villages,,Yellow jacket,AK,81335,USA,2853464789,Montezuma,lyndon.smitham@email.com,kgvmoxba,P,Y,UNK,N,Y,Any lab USA,BadCLIA,3279 Schroeder Mountain,,Yellow jacket,AK,81335,2365001476,Montezuma,see3r8,Any facility USA,35260 Dustin Crossroad,,Yellow jacket,AK,81335,2862149859,Montezuma,1368398388,Huels,Bradley,283 Runolfsson Drive,,Yellow jacket,AK,81335,2241497529,Montezuma,camp\n";
   *
   * export const file = (text: BlobPart) => {
   *   const blob = new Blob([text]);
   *   const file = new File([blob], "values.csv", { type: "text/csv" });
   *   File.prototype.text = jest.fn().mockResolvedValueOnce(text);
   *   return file;
   * };
   *
   * const validFile = () => file(validFileContents);
   */

  it("checks component with initial state", () => {
    const { container } = render(
      <SingleFileInput
        id="testId"
        name="testId"
        ariaLabel="upload a file"
        onChange={handleOnChange}
        required={false}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("checks component state when file has been selected", async () => {
    render(
      <SingleFileInput
        id="testId"
        name="testId"
        ariaLabel="upload a file"
        onChange={handleOnChange}
        required={false}
      />
    );
    const fileInput = screen.getByTestId("testId");
    await userEvent.upload(fileInput, file("test content"));
    expect(
      await screen.findByText(
        "Drag file here or choose from folder to change file"
      )
    );
    expect(handleOnChange).toHaveBeenCalledTimes(1);
  });

  it("checks component state when file is unselected", async () => {
    render(
      <SingleFileInput
        id="testId"
        name="testId"
        ariaLabel="upload a file"
        onChange={handleOnChange}
        required={false}
      />
    );
    const fileInput = screen.getByTestId("testId");
    fireEvent.change(fileInput, { target: { files: { length: 0 } } });
    expect(await screen.findByText("Drag file here or choose from folder"));
  });

  it("checks component state when file type is valid", async () => {
    render(
      <SingleFileInput
        id="testId"
        name="testId"
        ariaLabel="upload a file"
        onChange={handleOnChange}
        required={false}
        accept=".csv, text/csv"
      />
    );
    const fileInput = screen.getByTestId("testId");
    await userEvent.upload(fileInput, file("test content"));
    expect(
      await screen.findByText(
        "Drag file here or choose from folder to change file"
      )
    );
  });
  it("checks component state when file type is not valid", async () => {
    render(
      <SingleFileInput
        id="testId"
        name="testId"
        ariaLabel="upload a file"
        onChange={handleOnChange}
        required={false}
        accept=".csv, text/csv"
      />
    );
    const fileInput = screen.getByTestId("testId");

    let file = new File(["(⌐□_□)"], "chucknorris.png", { type: "image/png" });

    fireEvent.change(fileInput, {
      target: { files: { item: jest.fn().mockReturnValue(file), length: 1 } },
    });
    expect(await screen.findByText(/This is not a valid file type/i));
  });
});
