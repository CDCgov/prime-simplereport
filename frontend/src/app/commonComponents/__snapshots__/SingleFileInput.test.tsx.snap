// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single File Input checks component with initial state 1`] = `
<div>
  <div
    class="sr-single-file-input blank"
  >
    <input
      aria-describedby="file-input-specific-hint"
      aria-label="upload a file"
      class="usa-file-input"
      data-testid="testId"
      id="testId"
      name="testId"
      type="file"
    />
    <span
      id="file-input-specific-hint"
    >
      Drag file here or choose from folder
    </span>
  </div>
</div>
`;
